### PR TITLE
Elisa/5818 facility form

### LINKED ISSUE #5818: [Add Facility] Form not using proper element and error messages are not clear.
## Area or Component

Settings -> Add Facility page

## Quick Summary

The add facility form does is using a div element as a wrapper and the error messages for below fields do not provide specific instructions. 

- Facility Address [State]
- Ordering Provider [First name]
- Ordering Provider [Last name]
- Ordering Provider [Phone number]

### How to Replicate
Inspect the DOM and verify that the form tag is not being used.
Submit the form empty and check error messages for the fields listed above.

## Action Requested

Use the form element to wrap the form fields. Improve the error messaging for the inputs listed above.

## Screenshots (optional)

Screenshot 2023-05-16 at 1 49 39 PM
Screenshot 2023-05-16 at 1 49 06 PM
Screenshot 2023-05-16 at 1 48 48 PM
Screenshot 2023-05-16 at 1 40 28 PM

### PR DESCRIPTION
**Pull request by emyl3**
# FRONTEND PULL REQUEST

## Related Issue
Resolves #5818 

## Changes Proposed
- Fix accessibility concerns raised in #5818
- Refactor facility form to use `react-hook-form` library
- Remove `FacilitySchema.ts`

## Testing
- Ensure facilities can be added or edited
- Validation and error messages work correctly
  - particularly for edge cases:
    1. [states with unique CLIA numbers](https://github.com/CDCgov/prime-simplereport/blob/e86825cc30d7b3fc31275dd997b5723cf7f5f237/frontend/src/app/utils/clia.ts#L9)
    2. [states where ordering provider is not required](https://github.com/CDCgov/prime-simplereport/blob/e86825cc30d7b3fc31275dd997b5723cf7f5f237/frontend/src/config/constants.ts#L50): CO
    3. [states that cannot use SimpleReport](https://github.com/CDCgov/prime-simplereport/blob/e86825cc30d7b3fc31275dd997b5723cf7f5f237/frontend/src/config/constants.ts#LL8C1-L8C1)
- Addresses are being properly validated
- Axe dev tools scan passes
- on dev5 for testing

## Screenshots / Demos

- For large changes, please pair with a designer to ensure changes are as intended

### CODE REVIEW

File: frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx [hunk 1]
```diff
@@ -2,39 +2,33 @@ import React from "react";
...
-    updateProvider({ ...provider, [fieldName]: e.target.value });
-  };
+  const isRequired = requiresOrderProvider(
+    formCurrentValues.facility.state || ""
+  );
 
-  const { orderingProvider: provider } = facility;
-  const isRequired = requiresOrderProvider(facility.state || "");
+  const isValidNPI = (npi: string) => {
+    return /^\d{10}$/.test(npi);
```
Comment (emyl3) [inline frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx:27]: ⚠️ previously the regex to validate the NPI was [`/^\d{1,10}$/`](https://github.com/CDCgov/prime-simplereport/blob/97b716fcb0c73f982b9b35bc85621b536f73f7a1/frontend/src/app/Settings/Facility/facilitySchema.ts#LL35C6-L35C6), which means the NPI could be set to anything with less than 10 digits which according to our [error message](https://github.com/CDCgov/prime-simplereport/blob/97b716fcb0c73f982b9b35bc85621b536f73f7a1/frontend/src/app/Settings/Facility/facilitySchema.ts#LL45C11-L45C11) is incorrect.

Let me know if I should change this back to the previous regex.

File: frontend/src/app/Settings/Facility/FacilityForm.tsx [hunk 2]
```diff
@@ -226,16 +181,20 @@ const FacilityForm: React.FC<Props> = (props) => {
...
+    const originalFacilityAddress = getFacilityAddress(updatedValues.facility);
     const zipCodeData = await getZipCodeData(originalFacilityAddress.zipCode);
     const isValidZipForState = isValidZipCodeForState(
       originalFacilityAddress.state,
       zipCodeData
     );
 
     if (!isValidZipForState) {
-      showError("Invalid ZIP code for this state", "Form Errors");
+      setError("facility.zipCode", {
```
Comment (emyl3) [inline frontend/src/app/Settings/Facility/FacilityForm.tsx:193]: ⚠️ currently only validating if the facility's zip code and state combination is valid

we are not checking if the ordering provider's zip code and state (if provided, since it is not required) are valid combinations -- this is the current behavior as well

File: frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx [hunk 3]
```diff
@@ -105,66 +109,103 @@ const OrderingProvider: React.FC<Props> = ({
...
-          onChange={onChange}
+          value={formCurrentValues.orderingProvider.zipCode ?? undefined}
+          validationStatus={
+            errors?.orderingProvider?.zipCode?.type ? "error" : undefined
+          }
+          errorMessage={errors?.orderingProvider?.zipCode?.message}
+          registrationProps={register("orderingProvider.zipCode", {
+            required: false,
+            validate: {
+              validZip: (zipCode: string) =>
```
Comment (emyl3) [inline frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx:185]: ⚠️ decided against adding if the zip code was valid for the state validation here because we were making too many requests and getting back 429 errors from SmartyStreets 😅 -- so like it was done before, checking if the zip code and state combination are good only on submit after the form passes initial validation check

File: frontend/src/app/Settings/Facility/Components/FacilityInformation.tsx [hunk 4]
```diff
@@ -44,119 +54,166 @@ const FacilityInformation: React.FC<Props> = ({
...
+        />
+        <TextInput
+          label="Email"
+          name="email"
+          data-testid="facility-email"
+          value={formCurrentValues.facility?.email ?? undefined}
+          validationStatus={errors?.facility?.email?.type ? "error" : undefined}
+          errorMessage={errors?.facility?.email?.message}
+          registrationProps={register("facility.email", {
+            required: false,
```
Comment (BobanL) [inline frontend/src/app/Settings/Facility/Components/FacilityInformation.tsx:98]: Is required defaulted to false?

**Review by johanna-skylight — COMMENTED** [review]
Great job on the migration! I tested the form and it works smoothly. I left some questions but overall this looks awesome. Thanks for the work!

File: frontend/src/app/Settings/Facility/FacilityForm.tsx [hunk 5]
```diff
@@ -1,191 +1,146 @@
...
+    setFocus,
+    control,
+    setValue,
+    setError,
+    getValues,
+    getFieldState,
+    clearErrors,
+    formState: { errors, isSubmitting, isDirty },
+  } = useForm({
+    mode: "onBlur",
```
[the remark this answers is not in the document]
Comment (emyl3) [reply frontend/src/app/Settings/Facility/FacilityForm.tsx:117]: ooo ok! Yes, I will remove this line as I think the default onsubmit! Thank you!

File: frontend/src/app/Settings/Facility/Components/FacilityInformation.tsx (hunk 4 above)
Comment (emyl3) [reply frontend/src/app/Settings/Facility/Components/FacilityInformation.tsx:98]: ooo thank you!

File: frontend/src/app/Settings/Facility/Components/ManageDevices.tsx [hunk 6]
```diff
@@ -62,17 +44,18 @@ const ManageDevices: React.FC<Props> = ({
         <MultiSelect
           label="Device types"
           name="deviceTypes"
-          onChange={(newDeviceIds) => {
-            updateDevices(newDeviceIds);
-          }}
+          onChange={onChange}
```
[the remark this answers is not in the document]
Comment (emyl3) [reply frontend/src/app/Settings/Facility/Components/ManageDevices.tsx:47]: Yes! This onChange is being triggered to update the selected devices  and display an error if the device list is empty. Based on that, let me know if you see room for improvement! Thank you!!

**Review by johanna-skylight — DISMISSED** [review]
Took a closer look to manageDevices component and the onChange looks correct. Thanks Elisa!

**Review by BobanL — DISMISSED** [review]
Changes look good!

File: frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx [hunk 7]
```diff
@@ -105,66 +105,94 @@ const OrderingProvider: React.FC<Props> = ({
...
-          }}
+          value={formCurrentValues.orderingProvider.NPI ?? undefined}
+          registrationProps={register("orderingProvider.NPI", {
+            required: isRequired
+              ? "NPI should be a 10-digit numerical value (##########)"
+              : "",
+            validate: {
+              validNPI: (npi: string) =>
+                npi?.length
+                  ? isValidNPI(npi) ||
```
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx:116]: could we pull this string into some sort of shared variable with the above? don't want a typo / string drift to result in this check not returning the right value

File: frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx [hunk 8]
```diff
@@ -105,66 +105,94 @@ const OrderingProvider: React.FC<Props> = ({
...
           required={isRequired}
-          value={provider.NPI || ""}
-          onChange={onChange}
-          onBlur={() => {
-            validateField("orderingProvider.NPI");
-          }}
+          value={formCurrentValues.orderingProvider.NPI ?? undefined}
+          registrationProps={register("orderingProvider.NPI", {
+            required: isRequired
+              ? "NPI should be a 10-digit numerical value (##########)"
```
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx:111]: Could we pull these error messages into a shared constants file? Just in case we need to change the error status slightly in the future or if we do any checks against the error messages themselves, don't want to have potential typo / string drift problems. I'll leave a note to other instances as well just so you have an easy reference.

File: frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx [hunk 9]
```diff
@@ -105,66 +105,94 @@ const OrderingProvider: React.FC<Props> = ({
...
+            errors?.orderingProvider?.phone?.type ? "error" : undefined
           }
-          errorMessage={errors["orderingProvider.phone"]}
+          errorMessage={errors?.orderingProvider?.phone?.message}
+          registrationProps={register("orderingProvider.phone", {
+            required: isRequired ? "Ordering provider phone is required" : "",
+            validate: {
+              validPhone: (opPhone: string) =>
+                opPhone?.length
+                  ? phoneNumberIsValid(opPhone) ||
```
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx:140]: similar note to the comment [here](https://github.com/CDCgov/prime-simplereport/pull/5983/files#r1235287974)

File: frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx [hunk 10]
```diff
@@ -105,66 +105,94 @@ const OrderingProvider: React.FC<Props> = ({
...
+          validationStatus={
+            errors?.orderingProvider?.zipCode?.type ? "error" : undefined
+          }
+          errorMessage={errors?.orderingProvider?.zipCode?.message}
+          registrationProps={register("orderingProvider.zipCode", {
+            validate: {
+              validZip: (zipCode: string) =>
+                zipCode?.length
+                  ? zipCodeRegex.test(zipCode) ||
+                    "Ordering provider ZIP code is invalid"
```
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx:177]: similar note to the comment [here](https://github.com/CDCgov/prime-simplereport/pull/5983/files#r1235287974)

File: frontend/src/app/Settings/Facility/FacilityForm.test.tsx [hunk 11]
```diff
@@ -214,9 +214,9 @@ describe("FacilityForm", () => {
         exact: false,
       })[0];
       await act(async () => await userEvent.clear(facilityZipCodeInput));
-      await act(async () => await userEvent.tab());
+      await clickSaveButton();
       const facilityZipCodeWarning = await screen.findByText(
-        "Facility zip code is missing"
+        "Facility ZIP code is required"
```
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/FacilityForm.test.tsx:219]: similar note to the comment [here](https://github.com/CDCgov/prime-simplereport/pull/5983/files#r1235287974)

File: frontend/src/app/Settings/Facility/FacilityForm.tsx [hunk 12]
```diff
@@ -1,191 +1,145 @@
...
+  } = useForm({
+    defaultValues: getDefaultValues(facility),
+  });
+
+  const updateSelectedDevices = (selectedItems: string[]) => {
+    setValue("devices", selectedItems, { shouldDirty: true });
+    if (selectedItems.length === 0) {
+      setError("devices", {
+        type: "required",
+        message: "There must be at least one device",
```
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/FacilityForm.tsx:125]: similar note to the comment [here](https://github.com/CDCgov/prime-simplereport/pull/5983/files#r1235287974)

File: frontend/src/app/Settings/Facility/FacilityForm.tsx [hunk 13]
```diff
@@ -226,16 +180,20 @@ const FacilityForm: React.FC<Props> = (props) => {
...
     const isValidZipForState = isValidZipCodeForState(
       originalFacilityAddress.state,
       zipCodeData
     );
 
     if (!isValidZipForState) {
-      showError("Invalid ZIP code for this state", "Form Errors");
+      setError("facility.zipCode", {
+        type: "validZipForState",
+        message: "Invalid ZIP code for this state",
```
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/FacilityForm.tsx:194]: similar note to the comment [here](https://github.com/CDCgov/prime-simplereport/pull/5983/files#r1235287974)

File: frontend/src/app/Settings/Facility/FacilityForm.tsx [hunk 14]
```diff
@@ -286,37 +244,78 @@ const FacilityForm: React.FC<Props> = (props) => {
...
+    setValue("facility", facility);
+    setValue("orderingProvider", orderingProvider);
+    setValue("devices", devices);
+    clearErrors();
     await validateFacilityAddresses();
   };
 
+  const onError = () => {
+    showError(
+      "Please check the form to make sure you complete all of the required fields.",
```
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/FacilityForm.tsx:300]: similar note to the comment [here](https://github.com/CDCgov/prime-simplereport/pull/5983/files#r1235287974)

File: frontend/src/app/Settings/Facility/FacilityForm.tsx [hunk 15]
```diff
@@ -369,35 +367,46 @@ const FacilityForm: React.FC<Props> = (props) => {
...
+                inputTextRef: ref,
+                setFocus: () => setFocus(name),
+              }}
+              onChange={updateSelectedDevices}
+            />
+          )}
+          defaultValue={formCurrentValues.devices}
+          name="devices"
+          control={control}
+          rules={{ required: "There must be at least one device" }}
```
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/FacilityForm.tsx:402]: similar note to the comment [here](https://github.com/CDCgov/prime-simplereport/pull/5983/files#r1235287974)

File: frontend/src/app/Settings/Facility/Components/FacilityInformation.tsx [hunk 16]
```diff
@@ -44,119 +54,162 @@ const FacilityInformation: React.FC<Props> = ({
...
+          }
+          name="cliaNumber"
+          value={formCurrentValues.facility?.cliaNumber}
+          required
+          validationStatus={
+            errors?.facility?.cliaNumber?.type ? "error" : undefined
+          }
+          errorMessage={errors?.facility?.cliaNumber?.message}
+          registrationProps={register("facility.cliaNumber", {
+            required: "Facility CLIA number is required",
```
Comment (fzhao99) [inline frontend/src/app/Settings/Facility/Components/FacilityInformation.tsx:200]: Do we need to gate the boolean here based on whether CO is selected in the dropdown? I can't seem to submit a blank CLIA on the deployed instance when CO is selected. Lmk if I'm misunderstanding

https://github.com/CDCgov/prime-simplereport/assets/29645040/9d10dfcc-fce1-48c2-908c-70d7d79e38b6

**Review by fzhao99 — CHANGES_REQUESTED** [review]
did a follow up smoke test and found one potential issue with the CO edge case, so gating approval on that.

File: frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx (hunk 8 above)
Comment (emyl3) [reply frontend/src/app/Settings/Facility/Components/OrderingProvider.tsx:111]: Thank you, Bob! Will do!

File: frontend/src/app/Settings/Facility/Components/FacilityInformation.tsx [hunk 17]
```diff
@@ -44,119 +54,162 @@ const FacilityInformation: React.FC<Props> = ({
...
-        value={facility.cliaNumber}
-        required
-        onChange={onChange}
-        onBlur={() => {
-          validateField("cliaNumber");
-        }}
-        validationStatus={errors.cliaNumber ? "error" : undefined}
-        errorMessage={errors.cliaNumber}
-      />
+      <fieldset className="usa-fieldset">
```
[the remark this answers is not in the document]
Comment (emyl3) [reply frontend/src/app/Settings/Facility/Components/FacilityInformation.tsx:57]: It's what you pointed out as well as using the react-hook-form library so calling the [`register`](https://www.react-hook-form.com/api/useform/register/) method on each input and doing some validation there! Let me know if you have any more questions/concerns!

File: frontend/src/app/Settings/Facility/Components/FacilityInformation.tsx (hunk 16 above)
Comment (emyl3) [reply frontend/src/app/Settings/Facility/Components/FacilityInformation.tsx:200]: [Based on what was implemented previously](https://github.com/CDCgov/prime-simplereport/blob/04597c50f73cdd72a78442af636469ed98a0418f/frontend/src/app/Settings/Facility/facilitySchema.ts#L96), no matter what state we want require the CLIA number but for CO we don't validate it.

**Review by BobanL — APPROVED** [review]
👍

**Review by johanna-skylight — APPROVED** [review]

**Comment by emyl3** [comment]
@fzhao99 @BobanL @johanna-skylight -- so sorry everyone who reviewed! found one last bug where the suggested address wasn't showing up when provided with a ordering provider address -- added a test for this. Ready for (hopefully - final 🤞 😅) review!!
on dev5 for testing!
Screenshot 2023-06-22 at 11 41 56

**Review by fzhao99 — APPROVED** [review]
thank you for breaking out the strings!

**Review by BobanL — APPROVED** [review]
